drizzle/resample.py

```diff
@@ -0,0 +1,404 @@
+"""
+The `drizzle` module defines the `Drizzle` class, for combining input
+images into a single output image using the drizzle algorithm.
+"""
+import numpy as np
+
+from . import cdrizzle
+
+
+SUPPORTED_DRIZZLE_KERNELS = [
+    "square",
+    "gaussian",
+    "point",
+    "tophat",
+    "turbo",
+    "lanczos2",
+    "lanczos3"
+]
+
+CTX_PLANE_BITS = 32
+
+
+class Drizzle():
+    def __init__(self, n_images=1, kernel="square",
+                 fillval=None, out_shape=None, out_sci=None, out_wht=None,
+                 out_ctx=None):
+        """
+        n_images : int, optional
+            The number of images expected to be added to the resampled
+            output. When it is a positive number and ``out_ctx`` is `None`,
+            it allows to pre-allocate the necessary array for the output context
+            image. If the actual number of input images that will be resampled
+            will exceed initial allocation for the context image, additional
+            context planes will be added as needed (context array will "grow"
+            in the third dimention as new input images are added.)
+            This parameter is ignored when ``out_ctx`` is provided.
+
+        kernel: str, optional
+            The name of the kernel used to combine the input. The choice of
+            kernel controls the distribution of flux over the kernel. The kernel
+            names are: "square", "gaussian", "point", "tophat", "turbo",
+            "lanczos2", and "lanczos3". The square kernel is the default.
+
+        out_shape : tuple, None, optional
+            Shape (`numpy` order ``(Ny, Nx)``) of the output images (context
+            image will have a third dimension of size proportional to the number
+            of input images). This parameter is helpful when neither
+            ``out_sci``, ``out_wht``, nor ``out_ctx`` images are provided.
+
+        fillval: float, None, str, optional
+            The value of output pixels that did not have contributions from
+            input images' pixels. When ``fillval`` is either `None` or
+            ``"INDEF"`` and ``out_sci`` is provided, the values of ``out_sci``
+            will not be modified. When ``fillval`` is either `None` or
+            ``"INDEF"`` and ``out_sci`` is **not provided**, the values of
+            ``out_sci`` will be initialized to `numpy.nan`. If ``fillval``
+            is a string that can be converted to a number, then the output
+            pixels with no contributions from input images will be set to this
+            ``fillval`` value.
+
+        out_sci : 2d array of float32, optional
+            A 2D numpy array containing the output image produced by
+            drizzling. On the first call it should be set to zero.
+            Subsequent calls it will hold the intermediate results
```

Comment:
This is old documentation that applies to the "traditional" workflow in which we start with an image that is all `0`s or `NaN`s and keep adding inputs as needed. Initial  `out_sci` *does not* have to be all `0`s: it can be the previous "end result" upon which we can resample more inputs.